controllers/subscriptions.go

```diff
@@ -230,63 +228,83 @@ func GetOdfSubscription(cli client.Client) (*operatorv1alpha1.Subscription, erro
 }
 
-func GetVendorCsvNames(cli client.Client, kind odfv1alpha1.StorageKind) ([]string, error) {
+func GetCsvNames() []string {
 
-	var csvNames []string
-	var err error
-
-	if kind == FlashSystemKind {
-		csvNames = []string{IbmSubscriptionStartingCSV}
-	} else if kind == StorageClusterKind {
-		csvNames = []string{
-			OdfDepsSubscriptionStartingCSV,
-			OcsSubscriptionStartingCSV,
-			RookSubscriptionStartingCSV,
-			NoobaaSubscriptionStartingCSV,
-			PrometheusSubscriptionStartingCSV,
-			RecipeSubscriptionStartingCSV,
-			OcsClientSubscriptionStartingCSV,
-			CSIAddonsSubscriptionStartingCSV,
-			CephCSISubscriptionStartingCSV,
-		}
+	return []string{
+		OdfDepsSubscriptionStartingCSV,
+		OcsSubscriptionStartingCSV,
+		RookSubscriptionStartingCSV,
+		NoobaaSubscriptionStartingCSV,
+		PrometheusSubscriptionStartingCSV,
+		RecipeSubscriptionStartingCSV,
+		OcsClientSubscriptionStartingCSV,
+		CSIAddonsSubscriptionStartingCSV,
+		CephCSISubscriptionStartingCSV,
+		IbmSubscriptionStartingCSV,
 	}
-
-	return csvNames, err
 }
 
-func EnsureVendorCsv(cli client.Client, csvName string) (*operatorv1alpha1.ClusterServiceVersion, error) {
-
-	var err error
+func EnsureCsv(cli client.Client, csvName string) error {
 
 	csvObj := &operatorv1alpha1.ClusterServiceVersion{}
-	err = cli.Get(context.TODO(), types.NamespacedName{
-		Name: csvName, Namespace: OperatorNamespace}, csvObj)
-	if err != nil {
+	csvObj.Name, csvObj.Namespace = csvName, OperatorNamespace
+
+	if err := cli.Get(context.TODO(), client.ObjectKeyFromObject(csvObj), csvObj); err != nil {
 		if errors.IsNotFound(err) {
-			approvalErr := ApproveInstallPlanForCsv(cli, csvName)
-			if approvalErr != nil {
-				return nil, approvalErr
+			if present, err := isSubscriptionPresentForCsv(cli, csvName); err != nil {
+				return err
+			} else if present {
+				if err := ApproveInstallPlanForCsv(cli, csvName); err != nil {
+					return err
+				}
 			}
+			return nil
 		}
-		return nil, err
+		return err
 	}
-	_, err = controllerutil.CreateOrUpdate(context.TODO(), cli, csvObj, func() error {
-		csvObj.OwnerReferences = []metav1.OwnerReference{}
+
+	if _, err := controllerutil.CreateOrUpdate(context.TODO(), cli, csvObj, func() error {
 		return SetOdfSubControllerReference(cli, csvObj)
-	})
-	if err != nil && !errors.IsAlreadyExists(err) {
-		return nil, err
+	}); err != nil {
+		return err
 	}
 
 	isReady := csvObj.Status.Phase == operatorv1alpha1.CSVPhaseSucceeded &&
 		csvObj.Status.Reason == operatorv1alpha1.CSVReasonInstallSuccessful
 
 	if !isReady {
-		err = fmt.Errorf("CSV is not successfully installed")
-		return nil, err
+		return fmt.Errorf("CSV is not successfully installed")
 	}
 
-	return csvObj, err
+	return nil
+}
+
+func isSubscriptionPresentForCsv(cli client.Client, csvName string) (bool, error) {
+
+	// find the package name for the given csv
+	var pkgName string
+	subs := GetSubscriptions()
+	for _, sub := range subs {
+		if sub.Spec.StartingCSV == csvName {
```

Comment:
I am not sure you will always have a starting CSV. Also the starting CSV does not mean that actual CSV. 
An example: Assume the subscription starting CSV is pointing to an older CSV, but the install plan for that older CSV already got approved so the operator already moved to the next release (not the target release). In that case you will not find the starting CSV anymore